Assets/_BuildingGame/Scripts/Building System/BuildingBehaviour.cs

```diff
@@ -0,0 +1,81 @@
+using BuildingGame.Input;
+using System.Linq;
+using UnityEngine;
+
+namespace BuildingGame.BuildingSystem
+{
+    public class BuildingBehaviour : Core.BaseBehaviour
+    {
+        public InputActionsCore inputActions = new();
+
+        private IBuildingState _buildingState;
+        private IBuildingState _destroyState;
+
+        // Loads all structure scriptable objects
+        private Structure[] structures = Resources.LoadAll(
+            "ScriptableObjects/Structures", typeof(Structure)).Cast<Structure>().ToArray();
+
+        private int _currentSelectedStructureIndex = 0;
+
+
+        // field and property to manage current building state
+        private IBuildingState currentBuildingState;
+        private IBuildingState CurrentBuildingState
+        {
+            get { return currentBuildingState; }
+            set
+            {
+                currentBuildingState?.Exit(this);
+                currentBuildingState = value;
+                currentBuildingState.Start(this);
+            }
+        }
+
+        // field and property to manage current selected structure
+        private Structure currentSelectedStructure;
+        public Structure CurrentSelectedStructure
+        {
+            get { return currentSelectedStructure; }
+            set { currentSelectedStructure = value; }
+        }
+
+        public override void OnStart()
+        {
+            // Initialize all states
+            _buildingState = new BuildState();
+            _destroyState = new DestroyState();
+
+            inputActions.Enable();
+
+            // Select the first structure and start in building state
+            CurrentSelectedStructure = structures[_currentSelectedStructureIndex];
+            CurrentBuildingState = _buildingState;
+        }
+
+        public override void OnUpdate()
+        {
+            CurrentBuildingState?.Update(this);
+
+            // Switch states when the input has been pressed
+            if (inputActions.Building.SwitchState.WasPressedThisFrame())
```

Comment:
I would change these input checks to use the input events instead of checking every frame.